align/pnr/toplevel.py

```diff
@@ -251,44 +251,99 @@ def route( *, DB, idx, opath, adr_mode, PDN_mode, router_mode, selection=None):
 
     return results_name_map
 
-def place_and_route( *, DB, opath, fpath, numLayout, effort, adr_mode, PDN_mode, render_placements, verilog_d, router_mode, gui):
+def place_and_route( *, DB, opath, fpath, numLayout, effort, adr_mode, PDN_mode, verilog_d, router_mode, gui):
     TraverseOrder = DB.TraverseHierTree()
 
     for idx in TraverseOrder:
         place( DB=DB, opath=opath, fpath=fpath, numLayout=numLayout, effort=effort, idx=idx)
 
     idx = TraverseOrder[-1]
 
-    pairs = []
+    bboxes = []
+
 
     #
     # We need to make some changes if we want to just annotate a subhierarchy, for example if idx were not the toplevel
     # We need to search for the sub-hierarchies of that module and only retain those in our new verilog_d file
     # For visualizing primitives we need to use a different list (abstract and concrete template names)
     #
+    def r2wh( r):
+        return (r[2]-r[0], r[3]-r[1])
+
+    def gen_leaf_bbox_and_hovertext( ctn, p):
+        #return (p, list(gen_boxes_and_hovertext( placement_verilog_d, ctn)))
+        return (p, [ ((0, 0)+p, f'{ctn}<br>{0} {0} {p[0]} {p[1]}', True, 0)])
+
+    hack = []
+    hack2 = defaultdict(dict)
+
+    # Hack to get all the leaf cells sizes; still doesn't get the CC capacitors
+    for atn, gds_lst in DB.gdsData2.items():
+        ctns = [str(pathlib.Path(fn).stem) for fn in gds_lst]
+        for ctn in ctns:
+            if ctn in DB.lefData:
+                lef = DB.lefData[ctn][0]
+                p = lef.width, lef.height
+                if ctn in hack2[atn]:
+                    assert hack2[atn][ctn][0] == p
+                else:
+                    hack2[atn][ctn] = gen_leaf_bbox_and_hovertext( ctn, p)
+
+            else:
+                logger.error( f'LEF for concrete name {ctn} (of {atn}) missing.')
+
     for sel in range(DB.hierTree[idx].numPlacement):
         logger.info( f'DB.CheckoutHierNode( {idx}, {sel})')
         hN = DB.CheckoutHierNode( idx, sel)
         # create new verilog for each placement
         if verilog_d is not None:
             placement_verilog_d = gen_placement_verilog( hN, DB, verilog_d)
 
-            if render_placements:
-                dump_blocks2( placement_verilog_d, hN.name, sel, leaves_only=False, show=True)
+            logger.info( f"hpwl: {hN.HPWL}")
```

Comment:
@854768750 I'm expecting to see a HPWL value at this point corresponding to the placement I just checked out from the DB (idx, sel). I'm just getting -1 here.

Reply:
@854768750 Do we have to modify CheckoutHierNode as well?

Reply:
@stevenmburns I made some changed in https://github.com/ALIGN-analoglayout/ALIGN-public/pull/687/commits/517411ef60c6d293bde7ad0f4d72a30d72361ef8